python/ray/services.py

```diff
@@ -356,6 +356,143 @@ def check_version_info(redis_client):
+        # If a port is specified, then try only once to connect.
+        num_retries = 1
+    else:
+        port = new_port()
+    while counter < num_retries:
+        if counter > 0:
+            print("Redis failed to start, retrying now.")
+        p = subprocess.Popen([redis_filepath,
+                              "--port", str(port),
+                              "--loglevel", "warning",
+                              "--loadmodule", redis_module],
+                             stdout=stdout_file, stderr=stderr_file)
+        time.sleep(0.1)
+        # Check if Redis successfully started (or at least if it the executable
+        # did not exit within 0.1 seconds).
+        if p.poll() is None:
+            if cleanup:
+                all_processes[PROCESS_TYPE_REDIS_SERVER].append(p)
+            break
+        port = new_port()
+        counter += 1
+    if counter == num_retries:
+        raise Exception("Couldn't start credis module " + module + ".")
+
+    # Create a Redis client just for configuring Redis.
+    redis_client = redis.StrictRedis(host="127.0.0.1", port=port)
+    # Wait for the Redis server to start.
+    wait_for_redis_to_start("127.0.0.1", port)
+    # Configure Redis to generate keyspace notifications. TODO(rkn): Change
+    # this to only generate notifications for the export keys.
+    redis_client.config_set("notify-keyspace-events", "Kl")
+    # Configure Redis to not run in protected mode so that processes on other
+    # hosts can connect to it. TODO(rkn): Do this in a more secure way.
+    redis_client.config_set("protected-mode", "no")
+    # If redis_max_clients is provided, attempt to raise the number of maximum
+    # number of Redis clients.
+
+    # Increase the hard and soft limits for the redis client pubsub buffer to
+    # 128MB. This is a hack to make it less likely for pubsub messages to be
+    # dropped and for pubsub connections to therefore be killed.
+    cur_config = (redis_client.config_get("client-output-buffer-limit")
+                  ["client-output-buffer-limit"])
+    cur_config_list = cur_config.split()
+    assert len(cur_config_list) == 12
+    cur_config_list[8:] = ["pubsub", "134217728", "134217728", "60"]
+    redis_client.config_set("client-output-buffer-limit",
+                            " ".join(cur_config_list))
+    # Put a time stamp in Redis to indicate when it was started.
+    redis_client.set("redis_start_time", time.time())
+    # Record the log files in Redis.
+    record_log_files_in_redis(address(node_ip_address, port), node_ip_address,
+                              [stdout_file, stderr_file])
+    return port, p
+
+def start_credis(node_ip_address,
+                 port=None,
+                 redirect_output=False,
+                 cleanup=True):
+    """Start the credis global state store."""
+
+    stdout_file, stderr_file = new_log_files(
+        "credis_master", redirect_output)
+
+    master_port, _ = start_credis_component(
```

Comment:
These three `start_credis_component()` calls all pass `port` in, which is probably not what we want?